xmake/rules/c++/modules/modules_support/builder.lua

```diff
@@ -0,0 +1,370 @@
+        local build = _should_build(target, headerunit.path, bmifile, {key = key, headerunit = true})
+
+        if build then
+            _mark_build(target, key)
+        end
+
+        opt.build_headerunit(headerunit, key, bmifile, outputdir, build)
+    end
+end
+
+-- should we build this module or headerunit ?
+function _should_build(target, sourcefile, bmifile, opt)
+
+    -- force rebuild a module if any of its module dependency is rebuilt
+    local requires = opt.requires
+    if requires then
+        for required, _ in table.orderpairs(requires) do
+            local m = get_from_target_mapper(target, required)
+            if m then
+                local rebuild = compiler_support.memcache():get2("should_build_in" .. target:name(), m.key)
+                if rebuild then
+                    return true
+                end
+            end
+        end
+    end
+
+    -- or rebuild it if the file changed for headerunit and namedmodules
+    local objectfile = opt.objectfile
+    if compiler_support.has_module_extension(sourcefile) or (opt and opt.headerunit) then
+        local dryrun = option.get("dry-run")
+        local compinst = compiler.load("cxx", {target = target})
+        local compflags = compinst:compflags({sourcefile = sourcefile, target = target})
+
+        local dependfile = target:dependfile(bmifile or objectfile)
+        local dependinfo = target:is_rebuilt() and {} or (depend.load(dependfile) or {})
+
+        -- need build this object?
+        local depvalues = {compinst:program(), compflags}
+        local lastmtime = os.isfile(bmifile or objectfile) and os.mtime(dependfile) or 0
+
+        if dryrun or depend.is_changed(dependinfo, {lastmtime = lastmtime, values = depvalues}) then
+            return true
+        end
+    end
+
+    return false
+end
+
+-- generate meta module informations for package / other buildsystems import
+--
+-- e.g
+-- {
+--      "defines": ["FOO=BAR"]
+--      "imports": ["std", "bar"]
+--      "name": "foo"
+--      "file": "foo.cppm"
+-- }
+function _generate_meta_module_info(target, name, sourcefile, requires)
+
+    local modulehash = compiler_support.get_modulehash(target, sourcefile)
+    local module_metadata = {name = name, file = path.join(modulehash, path.filename(sourcefile))}
+
+    -- add definitions
+    module_metadata.defines = _builder(target).get_module_required_defines(target, sourcefile)
+
+    -- add imports
+    if requires then
+        for _name, _ in pairs(requires) do
+            module_metadata.imports = module_metadata.imports or {}
+            table.append(module_metadata.imports, _name)
+        end
+    end
+
+    return module_metadata
+end
+
+function _target_module_map_cachekey(target)
+    local mode = config.mode()
+    return target:name() .. "module_mapper" .. (mode or "")
+end
+
+function _is_duplicated_headerunit(target, key)
+    local mapper = get_target_module_mapper(target)
+    for _, mapped in pairs(mapper) do
```

Comment:
it maybe slow if mapper is too large.

or we can return mapper_keys as the second value then use `if mapper_keys[key] then`

```lua
local mapper, mapper_keys = get_target_module_mapper(target)
```

Reply:
like this?

```lua
-- get or create a target module mapper
function get_target_module_mapper(target)

    opt = opt or {}
    local memcache = compiler_support.memcache()
    local mapper = memcache:get2(target:name(), "module_mapper")
    if not mapper then
        mapper = {}
        memcache:set2(target:name(), "module_mapper", mapper)
    end

    return mapper, table.keys(mapper)
end

function _is_duplicated_headerunit(target, key)
    local mapper, mapper_keys = get_target_module_mapper(target)
    for _, mapped_key in ipairs(mapper_keys) do
        if mapped_key == key then
            return mapper[mapped_key]
        end
    end
end
```